### PR TITLE
Handle too long PR titles correctly

### LINKED ISSUE #16507: Pull request with long commit message in title fails to open
- Gitea version (or commit ref): try.gitea.io version (1.16.0+dev-28-g28f6f7bb0)
- Git version: 2.30.0.windows.2
- Operating system: Microsoft Windows 10 Pro, Version 10.0.19042 Build 19042
  
  
  
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes (https://try.gitea.io/ailuridae/my-test-repo/compare/master...feature/duh)
  - [ ] No
- Log gist:
Browser logs enumerated below.
    ```
  duh:1 Tracking Prevention blocked access to storage for https://secure.gravatar.com/avatar/e64fbf2d101290423cd7c42d02c378b0?d=identicon&s=96.
  duh:1 Tracking Prevention blocked access to storage for https://secure.gravatar.com/avatar/e64fbf2d101290423cd7c42d02c378b0?d=identicon&s=112.
  index.js?v=37edd3a08b9e064018380614f83a1b4b:62 jQuery.Deferred exception: SimpleMDE is not defined ReferenceError: SimpleMDE is not defined
      at Ba (https://try.gitea.io/assets/js/index.js?v=37edd3a08b9e064018380614f83a1b4b:60:1714)
      at Xm (https://try.gitea.io/assets/js/index.js?v=37edd3a08b9e064018380614f83a1b4b:29:33228)
      at https://try.gitea.io/assets/js/index.js?v=37edd3a08b9e064018380614f83a1b4b:62:20111
      at Generator.next (<anonymous>)
      at u (https://try.gitea.io/assets/js/index.js?v=37edd3a08b9e064018380614f83a1b4b:29:26053) undefined
  o.Deferred.exceptionHook @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  Pi @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  (anonymous) @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  Promise.then (async)
  Lt @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  Pi @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  setTimeout (async)
  (anonymous) @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  je @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  fireWith @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  fire @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  je @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  fireWith @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  ready @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  We @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  index.js?v=37edd3a08b9e064018380614f83a1b4b:60 Uncaught ReferenceError: SimpleMDE is not defined
      at Ba (index.js?v=37edd3a08b9e064018380614f83a1b4b:60)
      at Xm (index.js?v=37edd3a08b9e064018380614f83a1b4b:29)
      at index.js?v=37edd3a08b9e064018380614f83a1b4b:62
      at Generator.next (<anonymous>)
      at u (index.js?v=37edd3a08b9e064018380614f83a1b4b:29)
  Ba @ index.js?v=37edd3a08b9e064018380614f83a1b4b:60
  Xm @ index.js?v=37edd3a08b9e064018380614f83a1b4b:29
  (anonymous) @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  u @ index.js?v=37edd3a08b9e064018380614f83a1b4b:29
  setTimeout (async)
  o.readyException @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  (anonymous) @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  Lt @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  Pi @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  setTimeout (async)
  (anonymous) @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  je @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  fireWith @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  fire @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  je @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  fireWith @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  Pi @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  (anonymous) @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  Promise.then (async)
  Lt @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  Pi @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  setTimeout (async)
  (anonymous) @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  je @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  fireWith @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  fire @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  je @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  fireWith @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  ready @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  We @ index.js?v=37edd3a08b9e064018380614f83a1b4b:62
  ```

## Description
If you create a pull request from a commit with a very long commit message (>255 characters) and try to open a pull request without trimming down the title, opening the pull request will fail without any error notification. Typically, if you try to create a commit message longer than 100 characters in the UI, it will tell you to shorten the message. Furthermore, if you try to create a pull request with too long of a title, the input box will prevent you from typing more characters. However, this can be circumvented. This issue can be reproduced by:
1. Creating a commit with a very long message (>255 characters) on your local machine.
2. Pushing the commit to Gitea.
3. Creating a pull request with only that commit.
4. Selecting **Create Pull Request** on the New Pull Request page without editing the title.

I was able to replicate this issue on try.gitea.io, but I also can replicate it on my private Gitea instance running 1.13.2 with PostgreSQL. This issue arose when a developer on my team said that they could not open a pull request. We sat down and troubleshooted the issue, narrowing it down to this problem.

The only fix this really needs is some kind of error message that notifies the user that they need to trim down their PR title.

## Screenshots
https://user-images.githubusercontent.com/25494912/126497469-d65d819b-77f0-47a3-92c9-b3b38495aea8.mp4

### PR DESCRIPTION
**Pull request by zeripath**
The CompareAndPullRequestPost handler for POST to /compare
incorrectly handles returning errors to the user. For a start
it does not set the necessary markers to switch SimpleMDE
but it also does not immediately return to the form.

This PR fixes this by setting the appropriate values, fixing
the templates and preventing the suggestion of a too long
title.

Fix #16507

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: routers/web/repo/compare.go [hunk 1]
```diff
@@ -567,6 +567,15 @@ func PrepareCompareDiff(
 	} else {
 		title = headBranch
 	}
+	if len(title) > 255 {
+		if ctx.Data["content"] != nil {
+			ctx.Data["content"] = fmt.Sprintf("...%s\n\n%s", title[252:], ctx.Data["content"])
```
Comment (lafriks) [inline routers/web/repo/compare.go]: Could use Unicode character instead of "..." and use 254 characters for title

Comment (zeripath) [reply routers/web/repo/compare.go]: So I initially considered that but remembered - perhaps incorrectly - that github doesn't use … `U+2026` for some reason itself.

Comment (lafriks) [reply routers/web/repo/compare.go]: We use in all UI parts it already, on other hand this will go to database so idk 😄

Comment (zeripath) [reply routers/web/repo/compare.go]: I mean if we're willing to risk the mysql and mssql users' wrath and/or people who insist on non utf8 in their commit logs then switching to use ellipsis instead would be easy.

Comment (lafriks) [reply routers/web/repo/compare.go]: I don't understand this to be honest 😉... Not that there is still software that does not support utf8
Anyway this was just suggestion and is non blocking from my side

Comment (zeripath) [reply routers/web/repo/compare.go]: lets try the ellipsis.

File: routers/web/repo/pull.go [hunk 2]
```diff
@@ -1037,6 +1041,12 @@ func CompareAndPullRequestPost(ctx *context.Context) {
 			return
 		}
 
+		if len(form.Title) > 255 {
+			form.Content = "..." + form.Title[252:] + "\n\n" + form.Content
```
Comment (lafriks) [inline routers/web/repo/pull.go]: same here should be changed then

Comment (zeripath) [reply routers/web/repo/pull.go]:
```suggestion
			form.Content = "…" + form.Title[254:] + "\n\n" + form.Content
```

File: routers/web/repo/pull.go [hunk 3]
```diff
@@ -1037,6 +1041,12 @@ func CompareAndPullRequestPost(ctx *context.Context) {
 			return
 		}
 
+		if len(form.Title) > 255 {
+			form.Content = "…" + form.Title[254:] + "\n\n" + form.Content
+			form.Title = form.Title[:252] + "..."
```
Comment (zeripath) [inline routers/web/repo/pull.go]:
```suggestion
			form.Title = form.Title[:254] + "…"
```

**Review by lafriks — APPROVED** [review]

File: routers/web/repo/compare.go [hunk 4]
```diff
@@ -567,6 +567,15 @@ func PrepareCompareDiff(
 	} else {
 		title = headBranch
 	}
+	if len(title) > 255 {
+		if ctx.Data["content"] != nil {
+			ctx.Data["content"] = fmt.Sprintf("…%s\n\n%s", title[254:], ctx.Data["content"])
```
Comment (lunny) [inline routers/web/repo/compare.go]: This is still not correct because you have to consider utf8. Simply use `title[254:]` will break some of utf8 string.

**Review by lunny — CHANGES_REQUESTED** [review]

File: routers/web/repo/compare.go (hunk 4 above)
Comment (zeripath) [reply routers/web/repo/compare.go]: Yup looks like we'd need to cast to []rune and then drop the last rune.

Probably also need to consider if the last rune is a combining character too.

Comment (zeripath) [reply routers/web/repo/compare.go]: and we would need to consider this in chi-binding too.

Comment (zeripath) [reply routers/web/repo/compare.go]: I've switched the code to use byte length and truncation algorithm

**Review by techknowlogick — APPROVED** [review]

File: modules/util/truncate.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+// SplitStringAtByteN splits a string at byte n accounting for rune boundaries. (Combining characters are not accounted for.)
+func SplitStringAtByteN(input string, n int) (left, right string) {
+	if len(input) <= n {
+		left = input
+		return
+	}
+
+	if !utf8.ValidString(input) {
+		left = input[:n-3] + "..."
```
Comment (lunny) [inline modules/util/truncate.go:17]: `ValidString` don't mean it's totally ascii. Maybe we should correct it.

**Review by lunny — APPROVED** [review]

File: modules/util/truncate.go (hunk 5 above)
Comment (zeripath) [reply modules/util/truncate.go:17]: The test here is to simply abort trying to be nice if the string is not a valid utf8 string. (Which could be ASCII, Big5, UTF-16...)

If you don't have a utf-8 string in your commit header - getting it truncated nicely is going to be the least of your worries. We really don't handle other encodings in git commit headers.

**Comment by 6543** [comment]
@zeripath please backport :)

**Pull request merged**